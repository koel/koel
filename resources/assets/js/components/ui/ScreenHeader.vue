<template>
  <header class="screen-header" :class="layout">
    <aside class="thumbnail-wrapper">
      <slot name="thumbnail"></slot>
    </aside>

    <main>
      <div class="heading-wrapper">
        <h1 class="name">
          <slot></slot>
        </h1>
        <span class="meta text-secondary">
          <slot name="meta"></slot>
        </span>
      </div>

      <slot name="controls"></slot>
    </main>
  </header>
</template>

<script lang="ts" setup>
const props = withDefaults(defineProps<{ layout?: ScreenHeaderLayout }>(), { layout: 'expanded' })
</script>

<style lang="scss" scoped>
header.screen-header {
  --transition-duration: .3s;

  @media (prefers-reduced-motion) {
    --transition-duration: 0;
  }

  display: flex;
  align-items: flex-end;
  flex-shrink: 0;
  border-bottom: 1px solid var(--color-bg-secondary);
  position: relative;
  align-content: stretch;
  line-height: normal;
  padding: 1.8rem;

  &.expanded {
    .thumbnail-wrapper {
      margin-right: 1.5rem;
      width: 192px;

      > * {
        transform: scale(1);
      }
    }

    .meta {
      display: block;
    }

    main {
      flex-direction: column;
      align-items: flex-start;
    }
  }

  .thumbnail-wrapper {
    overflow: hidden;
    display: block;
    width: 0;
    transition: width var(--transition-duration);
    border-radius: 5px;

    > * {
      transform: scale(0);
      transform-origin: bottom left;
      transition: transform var(--transition-duration), width var(--transition-duration);
    }

    &:empty {
      display: none;
    }
  }

  main {
    flex: 1;
    display: flex;
    gap: 1.5rem;
    align-items: center;
    overflow: hidden;
  }

  h1.name {
    font-size: clamp(1.8rem, 3vw, 4rem);
    font-weight: var(--font-weight-bold);
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    margin-right: 1.5rem;
  }

  .heading-wrapper {
    width: 100%;
    overflow: hidden;
    flex: 1;
  }

  .meta {
    display: none;
    font-size: .9rem;
    line-height: 2;
    font-weight: var(--font-weight-light);

    a {
      color: var(--color-text-primary);

      &:hover {
        color: var(--color-highlight);
      }
    }

    > :slotted(* + *) {
      margin-left: .2rem;
      display: inline-block;

      &::before {
        content: 'â€¢';
        margin-right: .2rem;
        color: var(--color-text-secondary);
        font-weight: unset;
      }
    }
  }

  @media (max-width: 768px) {
    min-height: 0;

    .thumbnail-wrapper {
      display: none;
    }

    h1.name {
      font-weight: var(--font-weight-thin);
    }

    .meta {
      display: none;
    }
  }
}
</style>
