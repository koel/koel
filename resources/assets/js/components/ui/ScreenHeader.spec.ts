import { expect, it } from 'vitest'
import UnitTestCase from '@/__tests__/UnitTestCase'
import ScreenHeader from './ScreenHeader.vue'

new class extends UnitTestCase {
  protected test () {
    it('renders', () => {
      expect(this.render(ScreenHeader, {
        slots: {
          default: 'This Header',
          meta: '<p>Some meta</p>',
          controls: '<nav>Some controls</nav>',
          thumbnail: '<img src="https://placekitten.com/200/300" />',
        },
      }).html()).toMatchSnapshot()
    })
  }
}
