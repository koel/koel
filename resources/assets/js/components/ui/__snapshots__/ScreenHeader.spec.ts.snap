// Vitest Snapshot v1

exports[`renders 1`] = `
<header class="screen-header expanded" data-v-5691beb5="">
  <aside class="thumbnail-wrapper" data-v-5691beb5=""><img src="https://placekitten.com/200/300"></aside>
  <main data-v-5691beb5="">
    <div class="heading-wrapper" data-v-5691beb5="">
      <h1 class="name" data-v-5691beb5="">This Header</h1><span class="meta text-secondary" data-v-5691beb5=""><p>Some meta</p></span>
    </div>
    <nav>Some controls</nav>
  </main>
</header>
`;
